### PR TITLE
Timeline

### PR DESCRIPTION
**Pull request by goose-life**

### CODE REVIEW

**Comment by goose-life** [comment]
see #367

File: indigo_app/views/works.py [hunk 1]
```diff
@@ -56,6 +56,29 @@ def determine_place(self):
...
+        other_dates = [
+            ('assent_date', self.work.assent_date),
+            ('commencement_date', self.work.commencement_date),
+            ('repealed_date', self.work.repealed_date)
+        ]
+        # add to existing dates (e.g. if publication and commencement dates are the same)
+        for date in work_timeline:
+            for name, other_date in other_dates:
+                if date['date'] == other_date:
+                    date['{other_date}'.format(other_date=other_date)] = True
```
Comment (longhotsummer) [inline indigo_app/views/works.py]: This seems weird. What are you trying to do?

`str(other_date)` may be simpler than a format string, if that's what you want. Why do you want a date string to be True?

File: indigo_app/views/works.py [hunk 2]
```diff
@@ -56,6 +56,29 @@ def determine_place(self):
...
+
+        # add other dates to timeline
+        work_timeline = self.work.points_in_time()
+        other_dates = [
+            ('assent_date', self.work.assent_date),
+            ('commencement_date', self.work.commencement_date),
+            ('repealed_date', self.work.repealed_date)
+        ]
+        # add to existing dates (e.g. if publication and commencement dates are the same)
+        for date in work_timeline:
```
Comment (longhotsummer) [inline indigo_app/views/works.py]: Perhaps call this `entry` since it's not just a date.

File: indigo_app/views/works.py (hunk 1 above)
Comment (longhotsummer) [reply indigo_app/views/works.py]: I think you want `date[name] = True`L

File: indigo_app/views/works.py [hunk 3]
```diff
@@ -56,6 +56,29 @@ def determine_place(self):
...
+            for name, other_date in other_dates:
+                if date['date'] == other_date:
+                    date['{other_date}'.format(other_date=other_date)] = True
+        # add new dates (e.g. if assent is before any of the other dates)
+        existing_dates = [date['date'] for date in work_timeline]
+        for name, other_date in other_dates:
+            if other_date and other_date not in existing_dates:
+                work_timeline.append({
+                    'date': other_date,
+                    '{name}'.format(name=name): True,
```
Comment (longhotsummer) [inline indigo_app/views/works.py]: You can simply do `name: True`

File: indigo_app/views/works.py [hunk 4]
```diff
@@ -56,6 +56,49 @@ def determine_place(self):
...
+        other_dates = [
+            ('assent_date', self.work.assent_date),
+            ('commencement_date', self.work.commencement_date),
+            ('repealed_date', self.work.repealed_date)
+        ]
+        # add to existing dates (e.g. if publication and commencement dates are the same)
+        for entry in work_timeline:
+            for name, date in other_dates:
+                if entry['date'] == date:
+                    if name == 'commencement_date' and self.work.commencing_work:
```
Comment (longhotsummer) [inline indigo_app/views/works.py]: Minor style comment (totally optional):

You're effectively always setting `entry[name] = True`, so you could do that before line 71 and simplify the `if`s. Additionally, it's not going to hurt to have `event['commencing_work']` be None, so you can remove that check too.

File: indigo_app/views/works.py [hunk 5]
```diff
@@ -56,6 +56,49 @@ def determine_place(self):
...
+                    elif name == 'repealed_date' and self.work.repealed_by:
+                        entry['repealed_date'] = True
+                        entry['repealed_by'] = self.work.repealed_by
+                    else:
+                        entry[name] = True
+        # add new dates (e.g. if assent is before any of the other dates)
+        existing_dates = [entry['date'] for entry in work_timeline]
+        for name, date in other_dates:
+            if date and date not in existing_dates:
+                if name == 'commencement_date' and self.work.commencing_work:
```
Comment (longhotsummer) [inline indigo_app/views/works.py]: Similar comment from above applies here (but it's stylistic). You could start by doing everything in the final `else`, and then handle the exceptions.

Comment (goose-life) [reply indigo_app/views/works.py]: this way I don't have to then find the index of the dict in the list to update, so I was being a little lazy :grimacing:

File: indigo_app/templates/indigo_api/_work_points_in_time.html [hunk 6]
```diff
@@ -16,16 +16,46 @@
...
+        {% if event.assent_date %}
+          <h6>Assented to</h6>
+        {% endif %}
+
+        {% if event.initial %}
+          <h6>
+            Initial publication
+            {% if event.commencement_date %}
+              <span class="text-muted"> – Commenced on publication
+                {% if event.commencing_work %}
```
Comment (goose-life) [inline indigo_app/templates/indigo_api/_work_points_in_time.html]: actually, I don't think we need to pass this into the context separately – we can just use `if work.commencing_work`?

**Pull request merged**